controllers/propagator/replicatedpolicy_setup.go

```diff
@@ -1,48 +1,76 @@
 package propagator
 
 import (
+	"context"
+	"sync"
+	"time"
+
 	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	policiesv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
 	"open-cluster-management.io/governance-policy-propagator/controllers/common"
 )
 
-func (r *ReplicatedPolicyReconciler) SetupWithManager(mgr ctrl.Manager, additionalSources ...source.Source) error {
-	builder := ctrl.NewControllerManagedBy(mgr).
+func (r *ReplicatedPolicyReconciler) SetupWithManager(
+	mgr ctrl.Manager, dynWatcherSrc source.Source, updateSrc source.Source,
+) error {
+	return ctrl.NewControllerManagedBy(mgr).
 		Named("replicated-policy").
 		For(
 			&policiesv1.Policy{},
-			builder.WithPredicates(replicatedPolicyPredicates()),
-		)
-
-	for _, source := range additionalSources {
-		builder.WatchesRawSource(source, &handler.EnqueueRequestForObject{})
-	}
-
-	return builder.Complete(r)
+			builder.WithPredicates(replicatedPolicyPredicates(r.ResourceVersions))).
+		WatchesRawSource(
+			dynWatcherSrc,
+			// The dependency-watcher could create an event before the same sort of watch in the
+			// controller-runtime triggers an update in the cache. This tries to ensure the cache is
+			// updated before the reconcile is triggered.
+			&delayGeneric{
+				EventHandler: &handler.EnqueueRequestForObject{},
+				delay:        time.Second * 3,
+			}).
+		WatchesRawSource(
+			updateSrc,
+			&handler.EnqueueRequestForObject{}).
+		Complete(r)
 }
 
 // replicatedPolicyPredicates triggers reconciliation if the policy is a replicated policy, and is
-// not a pure status update.
-func replicatedPolicyPredicates() predicate.Funcs {
+// not a pure status update. It will use the ResourceVersions cache to try and skip events caused
+// by the replicated policy reconciler itself.
+func replicatedPolicyPredicates(resourceVersions *sync.Map) predicate.Funcs {
 	return predicate.Funcs{
 		CreateFunc: func(e event.CreateEvent) bool {
 			_, isReplicated := e.Object.GetLabels()[common.RootPolicyLabel]
+			if !isReplicated {
+				return false
+			}
 
-			// NOTE: can we ignore the create event from when this controller creates the resource?
-			return isReplicated
+			version := safeLoad(resourceVersions, e.Object)
```

Comment:
I think this logic would be better inside the reconciler itself instead of a predicate to fix the race condition you mentioned in the commit message.

In the reconciler, if the `resourceVersion` matches, then you can remove the entry in the map and return right away.

Reply:
What if the predicate for the root policy event source removed the replicated policy entry in `replicatedResourceVersions` before queuing the request? The same for the dependency watcher.

Then all reconcile requests from the root policy controller and the dependency watcher would be honored.

Reply:
That is a better alternative

Reply:
Actually, I think that might introduce a worse race condition (worse in the sense that it could cause a skipped reconcile, as opposed to the current problem which may result in an extra reconcile).

Example timeline, probably a worst-case:
1. A reconcile starts. It determines it needs to update the replicated policy. It begins its request to the API server.
2. An object in a hub-template is updated. A generic event is triggered, and gets to the predicate. The new generic predicate clears the resource version for the cache, and adds the reconcile to the queue.
3. The API server call returns the new resourceVersion. The reconciler stores it in its cache. **_This overrides what the generic predicate just did._**
4. The next reconcile is taken off the queue, and checks the resource version. It matches, so the reconcile is skipped.
5. The replicated policy now uses outdated information in its hub-template.

Reply:
Fair point, I think the cleanest solution is probably to split up this reconciler for one that handles the root policy and templating dependency watches event sources, and the other handling unexpected edits to the replicated policy.

Without that though, would this work considering that predicates are run before requests are deduplicated in the queue?

1. Have a `sync.Map` called `skipReconcile` which is a set of replicated policy namespace and names.
1. The replicated policy controller adds an entry whenever it knows it will generate a reconcile request to ignore.
1. The predicate on the replicated policy would continue the existing checks to ignore status updates and then checks the `skipReconcile` set to see if there is an entry for that replicated policy. If there is, it removes the entry and returns false. If there is not, it lets the replicated policy reconcile request through.
1. The other event sources don't have this predicate so it always lets the request through.

Reply:
I'll note that this assumes the predicate is run serially based on your previous comment. If it's not, then a `sync.Map` would be required for create, update, and deletes.

Reply:
I think that proposal would be less reliably correct. My feeling is that a "skip the next event" implementation could run into a case where it incorrectly skips something, as opposed to a "skip this specific event" implementation like this PR currently has.

If the controller sends an update on the Policy (which we want to skip), and then something else quickly updates it again (which we want to reconcile for), then the question is whether the controller-runtime guarantees that the predicate will be called twice. As I understand it, an etcd watch is the basis of these events - if multiple changes happen quickly, does etcd guarantee that every version in the middle is sent to the watcher? It feels like in the event of a networking issue, or if changes are sent to / come from different etcd replicas or something, then we might only be guaranteed to get the "latest" version of the resource, possibly skipping an intermediate event.

Splitting the reconciler again might be better.

Reply:
You should get a predicate call for each update since when a watch fails, it should be restarting from the last cached `resourceVersion`. To be safe, you could have the map be on the replicated policy + resource version like you have but just delete the map entry in the predicate if the stored resourceVersion matches or is less than the current resourceVersion.

Reply:
You should get a predicate call for each update since when a watch fails, it should be restarting from the last cached resourceVersion.

To be safer, we could use your current approach but for the create and update case, you could have a lock per replicated policy and start the lock before the create or update operation. Then unlock it after the new resourceVersion is written.

The predicate would obtain a read lock before seeing if the resource version matches.